hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java

```diff
@@ -222,4 +227,48 @@ protected HoodieTable getHoodieTable(HoodieWriteConfig writeConfig, HoodieTableM
   public BaseHoodieWriteClient<?, JavaRDD<HoodieRecord>, ?, ?> initializeWriteClient() {
     return new SparkRDDWriteClient(engineContext, metadataWriteConfig, true);
   }
+
+  @Override
+  protected HoodieData<HoodieRecord> getSecondaryIndexRecordsFromFileSlices(List<Pair<String, FileSlice>> partitionFileSlicePairs,
+                                                                            HoodieIndexDefinition indexDefinition,
+                                                                            int parallelism) throws IOException {
```

Comment:
Another choice is we add a `getEngineContext` abstract method for sub-classes, and these methods impl can move to the base class.

Reply:
Done.. added a `getEngineType` abstract method.